ksqldb-engine/src/main/java/io/confluent/ksql/planner/JoinTree.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.planner;
+
+import io.confluent.ksql.analyzer.Analysis.AliasedDataSource;
+import io.confluent.ksql.analyzer.Analysis.JoinInfo;
+import io.confluent.ksql.schema.ksql.FormatOptions;
+import io.confluent.ksql.util.KsqlException;
+import java.util.List;
+import java.util.Objects;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * {@code JoinTree} constructs the logical order for which the
+ * joins should be executed. At the moment, there is no optimization
+ * done and it simply follows the order of execution that the user
+ * indicates.
+ *
+ * <p>The algorithm is simple: the root is the very first source that
+ * we encounter (in the case of a single join, we ensure that the left
+ * root is the FROM source). From then on, any join that happens will
+ * check if either the left or right source is within the join tree and
+ * add the other to the corresponding side of the join tree.</p>
+ *
+ * <p>For example, take the following join statement:
+ * <pre>
+ * {@code
+ *    SELECT * FROM a
+ *      JOIN b ON a.id = b.id
+ *      JOIN c ON a.id = c.id;
+ * }
+ * </pre>
+ * The resulting join tree would look like:
+ * <pre>
+ * {@code
+ *       ⋈
+ *      / \
+ *     ⋈   C
+ *    / \
+ *   A   B
+ * }
+ * </pre>
+ * If the statement was modified so that the final expression was
+ * {@code JOIN c ON c.id = a.id} (note that {@code c.id} has moved
+ * to the left side of the join), then the resulting join tree would
+ * look like:
+ * <pre>
+ * {@code
+ *     ⋈
```

Comment:
I don't understand the rational behind this. Why create a right-deep tree here?

Reply:
this just let's the user express it however they'd like - it's not necessarily "right deep". If i write:
```sql
SELECT * FROM A 
  JOIN B ON b.id = a.id
  JOIN C on b.id = c.id
  JOIN D on d.id = b.id
```
I would get:
```
   ⋈
  / \
 D   ⋈
    / \
   ⋈   C
  / \
 B   A
```
The short of it is that I put the right expression on the right side and the left expression on the left side.

---

Really, it just made the code easier to write (I don't have to "flip" any of the expressions) and it gives more control to the user (which isn't a bad thing, given we don't do any optimizations)

Reply:
Nevermind... I just realized that these are the same trees. I Don't think it matters which source is on the left and which source is on the right in the join tree. I can rewrite it to always be left deep.